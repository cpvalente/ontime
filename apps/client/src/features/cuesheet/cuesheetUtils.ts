import { stringify } from 'csv-stringify/browser/esm/sync';
import { OntimeEntryCommonKeys, OntimeRundown, ProjectData, UserFields } from 'ontime-types';
import { millisToString } from 'ontime-utils';

/**
 * @description parses a field for export
 * @param {string} field
 * @param {*} data
 * @return {string}
 */

export const parseField = <T extends OntimeEntryCommonKeys>(field: T, data: unknown): string => {
  let val;
  switch (field) {
    case 'timeStart':
    case 'timeEnd':
      val = millisToString(data as number | null);
      break;
    case 'isPublic':
    case 'skip':
      val = data ? 'x' : '';
      break;
    default:
      val = data;
      break;
  }
  if (typeof data === 'undefined') {
    return '';
  }
  // all other values are strings
  return val as string;
};

/**
 * @description Creates an array of arrays usable by xlsx for export
 * @param {object} headerData
 * @param {array} rundown
 * @param {object} userFields
 * @return {(string[])[]}
 */
export const makeTable = (headerData: ProjectData, rundown: OntimeRundown, userFields: UserFields): string[][] => {
  const data = [
    ['Ontime Â· Schedule Template'],
    ['Project Title', headerData?.title || ''],
    ['Project Description', headerData?.description || ''],
    ['Public URL', headerData?.publicUrl || ''],
    ['Backstage URL', headerData?.backstageUrl || ''],
    [],
  ];

  const fieldOrder: OntimeEntryCommonKeys[] = [
    'timeStart',
    'timeEnd',
    'title',
    'presenter',
    'subtitle',
    'isPublic',
    'note',
    'colour',
    'endAction',
    'timerType',
    'skip',
    'user0',
    'user1',
    'user2',
    'user3',
    'user4',
    'user5',
    'user6',
    'user7',
    'user8',
    'user9',
  ];

  const fieldTitles = [
    'Time Start',
    'Time End',
    'Event Title',
    'Presenter Name',
    'Event Subtitle',
    'Is Public? (x)',
    'Note',
    'Colour',
    'End Action',
    'Timer Type',
    'Skip?',
  ];

  for (const field in userFields) {
    const fieldValue = userFields[field as keyof UserFields];
    const displayName = `${field}${fieldValue !== field && fieldValue !== '' ? `:${fieldValue}` : ''}`;
    fieldTitles.push(displayName);
  }

  data.push(fieldTitles);

  rundown.forEach((entry) => {
    const row: string[] = [];
    // @ts-expect-error -- not sure how to type this
    fieldOrder.forEach((field) => row.push(parseField(field, entry[field])));
    data.push(row);
  });

  return data;
};

/**
 * @description Converts an array of arrays to a csv file
 * @param {array[]} arrayOfArrays
 * @return {string}
 */
export const makeCSV = (arrayOfArrays: string[][]) => {
  const stringifiedData = stringify(arrayOfArrays);
  return stringifiedData;
};
