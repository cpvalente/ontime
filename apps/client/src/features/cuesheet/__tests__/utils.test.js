import { makeCSV, makeTable, parseField } from '../cuesheetUtils';

describe('parseField()', () => {
  it('returns a string from given millis on timeStart and TimeEnd', () => {
    const testData1 = 1000;
    const testData2 = 60000;
    expect(parseField('timeStart', testData1)).toBe('00:00:01');
    expect(parseField('timeEnd', testData2)).toBe('00:01:00');
    expect(parseField('timeEnd', testData2)).toBe('00:01:00');
  });

  describe('returns an x when isPublic is truthy, empty string otherwise', () => {
    const testTruthy = [1, true, 'x', 'test'];
    const testFalsy = ['', null, undefined, false, 0];

    testTruthy.forEach((value) => {
      test(`${value}`, () => {
        expect(parseField('isPublic', value)).toBe('x');
      });
    });
    testFalsy.forEach((value) => {
      test(`${value}`, () => {
        expect(parseField('isPublic', value)).toBe('');
      });
    });
  });

  it('returns an empty string on undefined fields', () => {
    expect(parseField('presenter')).toBe('');
  });

  describe('simply returns any other value in any other field', () => {
    const testFields = [
      { field: 'nothing', value: 123 },
      { field: 'title', value: 'test' },
      { field: 'presenter', value: 'test' },
      { field: 'subtitle', value: 'test' },
      { field: 'note', value: 'test' },
      { field: 'colour', value: 'test' },
      { field: 'user0', value: 'test' },
      { field: 'user1', value: 'test' },
      { field: 'user2', value: 'test' },
      { field: 'user3', value: 'test' },
      { field: 'user4', value: 'test' },
      { field: 'user5', value: 'test' },
      { field: 'user6', value: 'test' },
      { field: 'user7', value: 'test' },
      { field: 'user8', value: 'test' },
      { field: 'user9', value: 'test' },
    ];

    testFields.forEach((testCase) => {
      test(`${testCase.field}:${testCase.value}`, () => {
        expect(parseField(testCase.field, testCase.value)).toBe(testCase.value);
      });
    });
  });
});

describe('makeTable()', () => {
  it('returns array of arrays with given fields', () => {
    const headerData = {};
    const tableData = [
      {
        title: 'test title 1',
        presenter: '',
        timeStart: 0,
        timeEnd: 0,
        isPublic: 'x',
        user0: 'test',
        user1: 'test',
      },
    ];
    const userFields = {
      user0: 'test',
    };

    const table = makeTable(headerData, tableData, userFields);
    expect(table).toMatchSnapshot();
  });
});

describe('make CSV()', () => {
  it('joins an array of arrays with commas and newlines', () => {
    const testdata = [['field'], ['after newline', 'after comma'], ['', 'after empty']];
    expect(makeCSV(testdata)).toMatchInlineSnapshot(`
"field
after newline,after comma
,after empty
"
`);
  });
});
