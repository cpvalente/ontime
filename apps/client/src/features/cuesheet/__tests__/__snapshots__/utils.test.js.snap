// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`makeTable() > returns array of arrays with given fields 1`] = `
[
  [
    "Ontime Â· Schedule Template",
  ],
  [
    "Project Title",
    "",
  ],
  [
    "Project Description",
    "",
  ],
  [
    "Public URL",
    "",
  ],
  [
    "Backstage URL",
    "",
  ],
  [],
  [
    "Time Start",
    "Time End",
    "Event Title",
    "Presenter Name",
    "Event Subtitle",
    "Is Public? (x)",
    "Note",
    "Colour",
    "End Action",
    "Timer Type",
    "Skip?",
    "user0:test",
  ],
  [
    "00:00:00",
    "00:00:00",
    "test title 1",
    "",
    "",
    "x",
    "",
    "",
    "",
    "",
    "",
    "test",
    "test",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
]
`;
